components/script/dom/testbindingmaplike.rs

```diff
@@ -0,0 +1,90 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+// check-tidy: no specs after this line
+
+use crate::dom::bindings::cell::DomRefCell;
+use crate::dom::bindings::codegen::Bindings::TestBindingMaplikeBinding::TestBindingMaplikeMethods;
+use crate::dom::bindings::error::Fallible;
+use crate::dom::bindings::like::Maplike;
+use crate::dom::bindings::reflector::{reflect_dom_object_with_proto, Reflector};
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::globalscope::GlobalScope;
+use crate::maplike;
+use dom_struct::dom_struct;
+use indexmap::IndexMap;
+use js::rust::HandleObject;
+
+use super::bindings::error::Error;
+
+/// maplike<DOMString, long>
+#[dom_struct]
+pub struct TestBindingMaplike {
+    reflector: Reflector,
+    #[custom_trace]
+    internal: DomRefCell<IndexMap<DOMString, i32>>,
+}
+
+impl TestBindingMaplike {
+    fn new(global: &GlobalScope, proto: Option<HandleObject>) -> DomRoot<TestBindingMaplike> {
+        reflect_dom_object_with_proto(
+            Box::new(TestBindingMaplike {
+                reflector: Reflector::new(),
+                internal: DomRefCell::new(IndexMap::new()),
+            }),
+            global,
+            proto,
+        )
+    }
+
+    #[allow(non_snake_case)]
+    pub fn Constructor(
+        global: &GlobalScope,
+        proto: Option<HandleObject>,
+    ) -> Fallible<DomRoot<TestBindingMaplike>> {
+        Ok(TestBindingMaplike::new(global, proto))
+    }
+}
+
+impl TestBindingMaplikeMethods for TestBindingMaplike {
+    fn SetInternal(&self, key: DOMString, value: i32) {
+        self.internal.set(key, value)
+    }
+
+    fn ClearInternal(&self) {
+        self.internal.clear()
+    }
+
+    fn DeleteInternal(&self, key: DOMString) -> bool {
+        self.internal.delete(key)
+    }
+
+    fn HasInternal(&self, key: DOMString) -> bool {
+        self.internal.has(key)
+    }
+
+    fn GetInternal(&self, key: DOMString) -> Fallible<i32> {
+        // TODO: error type?
+        self.internal
+            .borrow()
+            .get(&key)
+            .ok_or_else(|| Error::Type(format!("No entry for key {key}")))
+            .copied()
+    }
+
+    fn Size(&self) -> u32 {
+        self.internal.size()
+    }
+}
+
+// this error is wrong because if we inline functions and then inline Self::Key all warnings are gone
+// TODO: FIX THIS
+#[allow(unrooted_must_root)]
```

Comment:
What error is being reported here? What does this comment mean?

Reply:
Well if I inline macro and remove allow I get this:
```console
error: Type must be rooted
  --> components/script/dom/testbindingmaplike.rs:90:34
   |
90 | fn get_index(&self,index:u32) -> Option<(Self::Key,Self::Value)>{
   |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[deny(unrooted_must_root)]` on by default

error: Type must be rooted
  --> components/script/dom/testbindingmaplike.rs:94:18
   |
94 | fn get(&self,key:Self::Key) -> Option<Self::Value>{
   |                  ^^^^^^^^^

error: Type must be rooted
  --> components/script/dom/testbindingmaplike.rs:94:32
   |
94 | fn get(&self,key:Self::Key) -> Option<Self::Value>{
   |                                ^^^^^^^^^^^^^^^^^^^

error: Type must be rooted
   --> components/script/dom/testbindingmaplike.rs:102:18
    |
102 | fn set(&self,key:Self::Key,value:Self::Value){
    |                  ^^^^^^^^^

error: Type must be rooted
   --> components/script/dom/testbindingmaplike.rs:102:34
    |
102 | fn set(&self,key:Self::Key,value:Self::Value){
    |                                  ^^^^^^^^^^^

error: Type must be rooted
   --> components/script/dom/testbindingmaplike.rs:106:18
    |
106 | fn has(&self,key:Self::Key) -> bool {
    |                  ^^^^^^^^^

error: Type must be rooted
   --> components/script/dom/testbindingmaplike.rs:114:21
    |
114 | fn delete(&self,key:Self::Key) -> bool {
    |                     ^^^^^^^^^

warning: `script` (lib) generated 1 warning
error: could not compile `script` due to 7 previous errors; 1 warning emitted
```

but if I replace all `Self::Key` with `DOMString` and `Self::Value` with `i32`, there is no error.

So something must be wrong with the compile plugin, probably due to Self::*.